hkmc2/shared/src/main/scala/hkmc2/semantics/Elaborator.scala

```diff
@@ -1322,6 +1337,131 @@ extends Importer:
           case L(d) => raise(d); go(tl, acc, ctx, flags)
       go(ps, Nil, ctx, ParamListFlags.empty)
   
+  def ident(id: Ident)(using Ctx): Ctxl[Opt[Term]] = ctx.get(id.name) match
+    case S(elem) => S(elem.ref(id))
+    case N =>
+      state.builtinOpsMap.get(id.name) match
+      case S(bi) => S(bi.ref(id))
+      case N => N
+  
+  def pattern(t: Tree): Ctxl[Pattern] =
+    import ucs.Desugarer.{Ctor, unapply}, Keyword.*, Pattern.*, InvalidReason.*
+    import ups.NaiveCompiler.isInvalidStringBounds, ucs.extractors.to
+    given TraceLogger = tl
+    /** Elaborate arrow patterns like `p => t`. Meanwhile, report all invalid
+     *  variables we found in `p`. */
+    def arrow(lhs: Tree, rhs: Tree): Ctxl[Pattern] =
+      val pattern = go(lhs)
+      val termCtx = ctx ++ pattern.variables.allocate
+      pattern.variables.report // Report all invalid variables we found in `pattern`.
+      Transform(pattern, term(rhs)(using termCtx))
+    /** Elaborate tuple patterns like `[p1, p2, ...ps, pn]`. */
+    def tuple(ts: Ls[Tree]): Ctxl[Pattern.Tuple] =
+      val z = (Ls[Pattern](), N: Opt[Pattern], Ls[Pattern]())
+      val (leading, spread, trailing) = ts.foldLeft(z):
+        case (acc @ (_, S(_), _), Spread(`...`, _, _)) =>
+          // Found two spreads in the same tuple pattern. Report an error.
+          raise(ErrorReport(msg"Multiple spread patterns are not supported." -> t.toLoc :: Nil))
+          acc // Do not modify the accumulator and skip this spread.
+        case ((leading, N, trailing), Spread(`...`, _, S(t))) =>
+          // Elaborate the spread pattern and add it to spread element.
+          (leading, S(go(t)), trailing)
+        case ((leading, N, trailing), Spread(`...`, _, N)) =>
+          // Empty spreads results in a wildcard pattern.
```

Comment:
Wait so `P(...)` and `P(_)` do the same thing? This is surprising. The former should match an arbitrary number of arguments.

Reply:
Sorry, what I actually meant is: if the `Spread` doesn’t have a `body` (i.e., `Spread(`...`, _, N)`), then _the sub-pattern matched against the spred scrutinee_ will be a wildcard.

```suggestion
          // If `Spread` does not have a `body`, then pattern for the spred
          // scrutinee will be a `Wildcard`.
```

Reply:
Please add a source code example to illustrate this situation. I don't remember what it means that a "spread does not have a body".

Reply:
It means a `Spread` that doesn’t spread anything. It’s just `...` literally.

Note that this is in the elaboration of tuple pattern. The enclosing `foldLeft` is accumulating three parts: the leading patterns, the spread pattern, and the trailing patterns. The comment says that if we found only `...` without a following pattern (i.e., the `body` field of `Spread` is set to `N`) then the spread pattern will be a wildcard, which means that we do not need to further match the middle part of the part.

Reply:
Improved in 9f4e6d920.